flink-core/src/main/java/org/apache/flink/api/common/eventtime/WatermarkStrategies.java

```diff
@@ -0,0 +1,230 @@
+
+/**
+ * WatermarkStrategies is a simply way to build a {@link WatermarkStrategy} by configuring
+ * common strategies.
+ */
+@PublicEvolving
+public final class WatermarkStrategies<T> {
+
+	/**
+	 * The {@link TimestampAssigner} to use. This can be {@code null} for cases where records come
+	 * out of a source with valid timestamps, for example from Kafka.
+	 */
+	@Nullable
+	private TimestampAssigner<T> timestampAssigner = null;
+
+	/** The base strategy for watermark generation. Starting point, is always set. */
+	private final WatermarkStrategy<T> baseStrategy;
+
+	/** Optional idle timeout for watermarks. */
+	@Nullable
+	private Duration idleTimeout;
+
+	private WatermarkStrategies(WatermarkStrategy<T> baseStrategy) {
+		this.baseStrategy = baseStrategy;
+	}
+
+	// ------------------------------------------------------------------------
+	//  builder methods
+	// ------------------------------------------------------------------------
+
+	/**
+	 * Add an idle timeout to the watermark strategy.
+	 * If no records flow in a partition of a stream for that amount of time, then that partition
+	 * is considered "idle" and will not hold back the progress of watermarks in downstream operators.
+	 *
+	 * <p>Idleness can be important if some partitions have little data and might not have events during
+	 * some periods. Without idleness, these streams can stall the overall event time progress of the
+	 * application.
+	 */
+	public WatermarkStrategies<T> withIdleness(Duration idleTimeout) {
+		checkNotNull(idleTimeout, "idleTimeout");
+		checkArgument(!(idleTimeout.isZero() || idleTimeout.isNegative()), "idleTimeout must be greater than zero");
+		this.idleTimeout = idleTimeout;
+		return this;
+	}
+
+	/**
+	 * Adds the given {@link TimestampAssigner} to this {@link WatermarkStrategies}.
+	 */
+	public <TA extends TimestampAssigner<T> & Serializable> WatermarkStrategies<T> withTimestampAssigner(TA timestampAssigner) {
+		checkNotNull(timestampAssigner, "timestampAssigner");
+		this.timestampAssigner = timestampAssigner;
+		return this;
+	}
+
+	/**
+	 * Adds the given {@link TimestampAssigner} to this {@link WatermarkStrategies}.
+	 */
+	public WatermarkStrategies<T> withTimestampAssigner(SerializableTimestampAssigner<T> timestampAssigner) {
+		checkNotNull(timestampAssigner, "timestampAssigner");
+		this.timestampAssigner = timestampAssigner;
+		return this;
+	}
+
+	/**
+	 * Build the watermark strategy.
+	 */
+	public WatermarkStrategy<T> build() {
+		WatermarkStrategy<T> strategy = this.baseStrategy;
+
+		if (idleTimeout != null) {
+			strategy = new WithIdlenessStrategy<>(strategy, idleTimeout);
+		}
+
+		if (timestampAssigner != null) {
+			strategy = new WithTimestampAssigner<>(strategy, timestampAssigner);
+		}
+
+		return strategy;
+	}
+
+	// ------------------------------------------------------------------------
+	//  builder entry points
+	// ------------------------------------------------------------------------
```

Comment:
nit: A matter of personal taste, but I prefer the entry methods to appear first in the class. That way the class reads more naturally from the top to the bottom.